src/haliax/nn/mup.py

```diff
@@ -0,0 +1,181 @@
+# Copyright 2025 The Levanter Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+
+import dataclasses
+import math
+import warnings
+from typing import Optional
+
+from jax.random import PRNGKey
+
+import haliax as hax
+
+from .._src.state_dict import (
+    Mod,
+    StateDict,
+    default_eqx_module_from_state_dict,
+    default_eqx_module_to_state_dict,
+    to_state_dict as _tree_to_state_dict,
+    with_prefix,
+)
+from ..axis import Axis, AxisSpec
+from ..jax_utils import named_call
+from ..quantization import DotGeneralOp
+
+from .embedding import Embedding
+from .linear import Linear
+
+
+class MupMixin:
```

Comment:
i kinda think this should be an object that is passed to mup-aware modules rather than inheritance. WDYT? like, Could we do

MupConfig(init_scale, lr_scale, active_scale)

I'd be fine with these infecting Linear and Embedding even, like

Linear:
   weight: NamedArray
   mup: MupConfig | None = None

Reply:
(this is more flexible and lets you do more ABC things)

Reply:
I think it might be helpful to see the coordinate check I just pushed to imagine how this might be used?

Reply:
More importantly, inheritance will come into play for optimizer transforms!

I want to have an operation which is map_mup_layers, similar to map_flattened_linear_layers (https://github.com/marin-community/levanter/blob/main/src/levanter/optim/util.py#L161-L211), which requires type checking